sources/models/scalar-derive/src/lib.rs

```diff
@@ -0,0 +1,493 @@
+struct MyWrapper {
+    some_field: WeirdType,
+}
+
+impl Validate for MyWrapper {
+    fn validate<T: Into<WeirdType>>(input: T) -> Result<MyWrapper, ValidationError> {
+        Ok(Self{ some_field: input.into() })
+    }
+}
+```
+
+*/
+
+use darling::FromAttributes;
+use proc_macro::TokenStream;
+use quote::{format_ident, ToTokens};
+use quote::{quote, TokenStreamExt};
+use syn::__private::TokenStream2;
+use syn::{parse_macro_input, Data, DataStruct, DeriveInput, Fields};
+
+/// A convenience macro for implementing Bottlerocket model types. See description in the lib
+/// documentation or README.
+#[proc_macro_derive(Scalar, attributes(scalar))]
+pub fn scalar(input: TokenStream) -> TokenStream {
+    // Parse the input tokens.
+    let derive_input = parse_macro_input!(input as DeriveInput);
+    let settings = RawSettings::from_attributes(derive_input.attrs.as_slice())
+        .expect("Unable to parse `scalar` macro arguments");
+
+    // Further parse the input.
+    let struct_info = StructInfo::new(&derive_input, settings);
+
+    // Write impls.
+    let mut ast2 = TokenStream2::new();
+    struct_info.write_impls(&mut ast2);
+    ast2.into_token_stream().into()
+}
+
+/// Store args given by the user inside `#[scalar(as_ref_str=false, ...)]` using `darling` (it's
+/// kind of like `clap` for derive macro arguments).
+#[derive(Debug, Clone, Default, FromAttributes)]
+#[darling(attributes(scalar))]
+#[darling(default)]
+struct RawSettings {
+    /// Whether or not `AsRef<str>` (and similar impls) should be created. If the inner type is
+    /// `String` (or implements `AsRef<str>`) then it is more convenient when using the `Scalar` if
+    /// it treats references to its inner type as `&str` instead of `&String`. This defaults to
+    /// `true` when the inner type is `String` and `false` otherwise.
+    as_ref_str: Option<bool>,
+    /// The name of the field that holds the inner value in the struct. Defaults to "inner".
+    inner: Option<String>,
+}
+
+/// Once we parse the incoming AST and see what our struct is named, see what its inner type is,
+/// and introspect the macro arguments, we save the resultant information in this structure for
+/// use during code generation.
+#[derive(Debug, Clone)]
+struct StructInfo {
+    /// The typename of the struct, that the `Scalar` derive macro was called on.
+    scalar: String,
+    /// The name of the field, inside the `scalar` struct, that holds the "inner" value.
+    inner_field: String,
+    /// The type of the `inner_field`.
+    inner_type: String,
+    /// Whether or not we should treat the inner reference type as `&str`.
+    as_ref_str: bool,
+}
+
+impl StructInfo {
+    fn new(derive_input: &DeriveInput, settings: RawSettings) -> Self {
+        let scalar = derive_input.ident.to_string();
+
+        let (inner_field, inner_type) = match derive_input.data.clone() {
+            Data::Struct(s) => find_inner_field(s, settings.inner.as_ref().map(|s| s.as_str())),
```

Comment:
While reading this I wondered whether or not it is right to allow a struct with more than one field. Is that really a Scalar?

I think though that e.g. a URI which could be split to store the scheme, domain, and path under separate properties would still find this useful. Probably best not to dwell.